src/partial-assembly.md

```diff
@@ -0,0 +1,171 @@
+
+This sequence of operations highlights an inherent hierarchical structure
+present in all finite element operators where the evaluation starts on *global
+(trial) degrees of freedom (dofs) on the whole mesh*, restricts to *degrees of
+freedom on subdomains* (groups of elements), then moves to independent *degrees
+of freedom on each element*, transitions to independent *quadrature points* in
+reference space, performs the integration, and then goes back in reverse order
+to global (test) degrees of freedom on the whole mesh.
+
+This is illustrated below for the case of a symmetric linear operator. We use
+the notions **T-vector**, **L-vector**, **E-vector** and **Q-vector** to
+represent the sets corresponding to the (true) degrees of freedom on the global
+mesh, the split local degrees of freedom on the subdomains, the split degrees
+of freedom on the mesh elements, and the values at quadrature points,
+respectively.
+
+We refer to the operators that connect the different types of vectors as:
+
+- Subdomain restriction **P**
+- Element restriction **G**
+- Basis (Dofs-to-Qpts) evaluator **B**
+- Operator at quadrature points **D**
+
+More generally, when the test and trial space differ, they get their own
+versions of **P**, **G** and **B**.
+
+![Operator Decomposition](img/libceed.png "Operator Decomposition")
+
+Note that in the case of adaptive mesh refinement (AMR), the restrictions **P**
+and **G** will involve not just extracting sub-vectors, but evaluating values
+at constrained degrees of freedom through the AMR interpolation. There can also
+be several levels of subdomains (**P1**, **P2**, etc.), and it may be
+convenient to split **D** as the product of several operators (**D1**, **D2**,
+etc.).
+
+## Partial Assembly in MFEM
+
+Since the global operator **A** is just a series of variational restrictions
+with **B**, **G** and **P**, starting from its point-wise kernel **D**, a
+"matrix-vector product" with **A** can be performed by evaluating and storing
+some of the innermost variational restriction matrices, and applying the rest
+of the operators "on-the-fly". For example, one can compute and store a global
+matrix on T-vector level. Alternatively, one can compute and store only the
+subdomain (L-vector) or element (E-vector) matrices and perform the action of
+**A** using matvecs with **P** or **P** and **G**. While these options are
+natural for low-order discretizations, they are not a good fit for high-order
+methods due to the amount of FLOPs needed for their evaluation, as well as the
+memory transfer needed for a matvec.
+
+MFEM's **partial assembly** functionality computes and stores only **D** (or
+portions of it) and evaluates the actions of **P**, **G** and **B** on-the-fly.
+Critically for performance, MFEM takes advantage of the tensor-product
+structure of the degrees of freedom and quadrature points on quadrilateral and
+hexahedral elements to perform the action of **B** without storing it as a
+matrix.
```

Comment:
Another point that may be worth mentioning here is that `B` is element-wise block diagonal where the blocks only change when either the order or the reference geometry of the finite element change. Currently, MFEM supports only the case of fixed order + same geometry, where all blocks of `B` are the same.

Reply:
Good point. I added a sentence here.